### PR TITLE
WebServer and WebClient race conditions fixed

### PR DESCRIPTION
**Pull request by Verdent**
Signed-off-by: David Kral <david.k.kral@oracle.com>

### CODE REVIEW

File: webserver/webserver/src/main/java/io/helidon/webserver/ForwardingHandler.java [hunk 1]
```diff
@@ -238,31 +243,33 @@ protected void channelRead0(ChannelHandlerContext ctx, Object msg) {
                 prevRequestFuture = null;
             }
 
+            requestDrained = new CompletableFuture<>();
             // Create response and handler for its completion
             BareResponseImpl bareResponse =
-                    new BareResponseImpl(ctx, request, publisher::isCompleted, prevRequestFuture, requestId);
+                    new BareResponseImpl(ctx, request, publisher::isCompleted, publisher::hasRequests, prevRequestFuture,
```
[the remark this answers is not in the document]
Comment (Verdent) [reply webserver/webserver/src/main/java/io/helidon/webserver/ForwardingHandler.java:249]: Oh I see now where this comes from. IDEA reformatted those automatically for me while I was restoring this from backup after accidental erasure. It is bound to more then this file. I will revert any irrelevant changes.

**Comment by Verdent** [comment]
Fixes #3242

File: webclient/webclient/src/main/java/io/helidon/webclient/RequestContentSubscriber.java [hunk 2]
```diff
@@ -145,19 +146,24 @@ public void onComplete() {
...
+                    if (future.isSuccess()) {
+                        sent.complete(serviceRequest);
+                        System.out.println("Client: Full entity sent, Thread: " + Thread.currentThread().getName());
+                        LOGGER.finest(() -> "(client reqID: " + requestId + ") Request sent");
+                    }
+                });
     }
 
     private void sendData(DataChunk data) {
+//        System.out.println("Sending data");
```
Comment (tomas-langer) [inline webclient/webclient/src/main/java/io/helidon/webclient/RequestContentSubscriber.java:166]: Please remove

File: webclient/webclient/src/main/java/io/helidon/webclient/WebClientRequestBuilderImpl.java [hunk 3]
```diff
@@ -95,6 +95,11 @@
     static final AttributeKey<ConnectionIdent> CONNECTION_IDENT = AttributeKey.valueOf("connectionIdent");
     static final AttributeKey<Long> REQUEST_ID = AttributeKey.valueOf("requestID");
 
+    /**
+     * Whether the channel will be closed and keep-alive caching should not be applied.
+     */
+    static final AttributeKey<Boolean> WILL_CLOSE = AttributeKey.valueOf("willClose");
```
Comment (tomas-langer) [inline webclient/webclient/src/main/java/io/helidon/webclient/WebClientRequestBuilderImpl.java:101]: Shouldn't we use some kind of Helidon specific prefix for attributes? Are there any reserved names by Netty?

File: webclient/webclient/src/main/java/io/helidon/webclient/RequestContentSubscriber.java [hunk 4]
```diff
@@ -145,19 +146,24 @@ public void onComplete() {
...
-
-        WebClientRequestImpl clientRequest = channel.attr(REQUEST).get();
-        WebClientServiceRequest serviceRequest = clientRequest.configuration().clientServiceRequest();
-        sent.complete(serviceRequest);
-        LOGGER.finest(() -> "(client reqID: " + requestId + ") Request sent");
+                .addListener(ChannelFutureListener.CLOSE_ON_FAILURE)
+                .addListener(future -> {
+                    if (future.isSuccess()) {
+                        sent.complete(serviceRequest);
+                        System.out.println("Client: Full entity sent, Thread: " + Thread.currentThread().getName());
```
Comment (tomas-langer) [inline webclient/webclient/src/main/java/io/helidon/webclient/RequestContentSubscriber.java:159]: Please remove

**Review by tomas-langer — CHANGES_REQUESTED** [review]

File: webclient/webclient/src/main/java/io/helidon/webclient/WebClientRequestBuilderImpl.java (hunk 3 above)
Comment (Verdent) [reply webclient/webclient/src/main/java/io/helidon/webclient/WebClientRequestBuilderImpl.java:101]: I don't think there are any to be fair. Or at least I haven't heard of any. I can check if you want.

File: webclient/webclient/src/main/java/io/helidon/webclient/RequestContentSubscriber.java (hunk 4 above)
Comment (Verdent) [reply webclient/webclient/src/main/java/io/helidon/webclient/RequestContentSubscriber.java]: Removed. Thank you

File: webclient/webclient/src/main/java/io/helidon/webclient/RequestContentSubscriber.java (hunk 2 above)
Comment (Verdent) [reply webclient/webclient/src/main/java/io/helidon/webclient/RequestContentSubscriber.java]: Removed. Thank you

**Review by tomas-langer — APPROVED** [review]
LGTM

**Pull request merged**